core/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java

```diff
@@ -174,10 +175,13 @@ public MetaData applyChanges(MetaData oldMetaData, RoutingTable newRoutingTable)
             // Prevent set of inSyncAllocationIds to grow unboundedly. This can happen for example if we don't write to a primary
             // but repeatedly shut down nodes that have active replicas.
             // We use number_of_replicas + 1 (= possible active shard copies) to bound the inSyncAllocationIds set
+            // Only trim the set of allocation ids when it grows, otherwise we might trim too eagerly when the number
+            // of replicas was decreased while shards were unassigned.
```

Comment:
the in-sync set is managed on a per-shard id (index name + shard number) basis so that scenario does not apply.